src/v/kafka/server/handlers/produce.cc

```diff
@@ -279,37 +279,63 @@ static partition_produce_stages produce_topic_partition(
+                                batch.key_offsets()->push_back(
+                                  model::key_offset(
+                                    r.offset_delta(),
+                                    r.key_size(),
+                                    iobuf_to_bytes(r.key())));
                             });
-                          return;
-                      }
-                      (void)ss::smp::submit_to(
-                        source_shard,
-                        [dispatch = std::move(dispatch)]() mutable {
-                            dispatch->set_value();
-                            dispatch.reset();
-                        });
-                  })
-                  .then([f = std::move(stages.produced)]() mutable {
-                      return std::move(f);
-                  });
+                      });
+                }
+                return f.then([partition = std::move(partition),
+                               batch = std::move(batch),
+                               ntp = std::move(ntp),
+                               dispatch = std::move(dispatch),
+                               num_records,
+                               bid,
+                               acks,
+                               source_shard]() mutable {
+                    auto reader = reader_from_lcore_batch(std::move(batch));
+                    auto stages = partition_append(
+                      ntp.tp.partition,
+                      ss::make_lw_shared<replicated_partition>(
+                        std::move(partition)),
+                      bid,
+                      std::move(reader),
+                      acks,
+                      num_records);
+                    return stages.dispatched
+                      .then_wrapped(
+                        [source_shard, dispatch = std::move(dispatch)](
+                          ss::future<> f) mutable {
+                            if (f.failed()) {
+                                (void)ss::smp::submit_to(
+                                  source_shard,
+                                  [dispatch = std::move(dispatch),
+                                   e = f.get_exception()]() mutable {
+                                      dispatch->set_exception(e);
+                                      dispatch.reset();
+                                  });
+                                return;
+                            }
+                            (void)ss::smp::submit_to(
+                              source_shard,
+                              [dispatch = std::move(dispatch)]() mutable {
+                                  dispatch->set_value();
+                                  dispatch.reset();
+                              });
+                        })
+                      .then([f = std::move(stages.produced)]() mutable {
+                          return std::move(f);
+                      });
+                });
```

Comment:
This part did not change, i think this would befit if we would split formatting and code changes.